### PR TITLE
Add choice observers in the ledger model.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
To keep the distinction between "observer" and "choice observer"
clear I'm trying to always refer to the former as "contract observer".

The section on divulgence should probably be reworked to favor choice
observers, at some point, but I'm not sure how to do that.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (cocreature) [inline docs/source/concepts/ledger-model/ledger-privacy.rst:46]:
```suggestion
Choice observers get to see a all exercises of a specific choice on a contract, and to view its consequences.
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice.

**Review by cocreature — APPROVED** [review]
Nice thanks!

**Comment by cocreature** [comment]
cc @andreaslochbihler-da @oggy- since this changes the ledger model.

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (oggy-) [inline docs/source/concepts/ledger-model/ledger-privacy.rst:45]:
```suggestion
In addition to contract observers, the contract model can also specify **choice observers** on individual **Exercise** actions.
```

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (oggy-) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:46]: Right now the ledger model doesn't include a notion of a choice. I do want to add it when I find a bit of spare time, but until then I would maybe not mention them here.

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (oggy-) [inline docs/source/concepts/ledger-model/ledger-privacy.rst:102]: Since the whole list is a distinction, I'd suggest just adding an item for choice observers rather than duplicating and including nested disjunctions

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (oggy-) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:46]: A short blurb choice observers should also be added to the "DAML: Defining Contract Models Compactly" section

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (sofiafaro-da) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:102]: That's a great idea! 👍

**Comment by oggy-** [comment]
Looks good overall, thanks for the write-up. Please consider my comments before merging and ping if unclear.

**Review by oggy- — APPROVED** [review]

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (sofiafaro-da) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:45]: Thanks!

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (sofiafaro-da) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:46]: @cocreature  I think a choice observer can depend on the arguments to the choice, right? So they really are observers to a specific exercise, not to every exercise on the choice.

Comment (cocreature) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:46]: good point, I guess it might still be useful to state somewhere that they are defined by choice but that brings up to @oggy-’s point that choices aren’t included in here at all so keeping it at exercises should be perfectly fine.

File: docs/source/concepts/ledger-model/ledger-daml.rst
Comment (sofiafaro-da) [inline docs/source/concepts/ledger-model/ledger-daml.rst:36]: @oggy-  I added this line here in the "Daml: Defining Contract Models Compactly" page. (Not going into details on the DAML syntax for choice observers here, since that's a separate part of the docs.)

File: docs/source/concepts/ledger-model/ledger-privacy.rst
Comment (bame-da) [reply docs/source/concepts/ledger-model/ledger-privacy.rst:46]: Maybe we should have called them "exercise observer"... In any case I think @sofiafaro-da 's phrasing was good.

**Pull request merged**
